certstore_linux.go

```diff
@@ -1,14 +1,369 @@
+		copy(T[len(oid):], digest)
+		digest = T
+	}
+	sd := C.SECITEM_AllocItem(nil, nil, C.uint(C.PK11_SignatureLen(key)))
+	hashed := C.SECITEM_AllocItem(nil, nil, C.uint(len(digest)))
+	if hashed == nil {
+		return nil, errors.New("SECITEM_AllocItem failed")
+	}
+	C.memcpy(unsafe.Pointer(hashed.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+	var mechanism C.ulong
+	switch key.keyType {
+	case C.rsaKey:
+		switch hash {
+		/*
+		case crypto.SHA1:
+			mechanism = C.CKM_SHA1_RSA_PKCS
+		case crypto.SHA256:
+			mechanism = C.CKM_SHA256_RSA_PKCS
+		case crypto.SHA384:
+			mechanism = C.CKM_SHA384_RSA_PKCS
+		case crypto.SHA512:
+			mechanism = C.CKM_SHA512_RSA_PKCS
+		*/
+		default:
+			mechanism = C.CKM_RSA_PKCS
+		}
+	case C.ecKey:
+		switch hash {
+		/*
+		case crypto.SHA1:
+			mechanism = C.CKM_ECDSA_SHA1
+		*/
+		default:
+			mechanism = C.CKM_ECDSA
+		}
+	default:
+		return nil, errors.New(fmt.Sprintf("Unknown key type: %d", int(key.keyType)))
+	}
+	ret := C.PK11_SignWithMechanism(key, mechanism, nil, sd, hashed)
+	if ret != 0 {
+		return nil, errors.New("could not sign")
+	}
+	var sig = C.GoBytes(unsafe.Pointer(sd.data), C.int(sd.len))
+	if key.keyType == C.ecKey {
+		if len(sig)%2 != 0 {
+			return nil, errors.New("bad ecdsa signature")
+		}
+		type ecdsaSignature struct {
+			R, S *big.Int
+		}
+		r := new(big.Int).SetBytes(sig[:len(sig)/2])
+		s := new(big.Int).SetBytes(sig[len(sig)/2:])
+		asn, err := asn1.Marshal(ecdsaSignature{r, s})
+		if err != nil {
+			return nil, err
+		}
+		sig = asn
+	}
+	return sig, nil
+}
+
+func newNssIdentity(node *C.CERTCertListNode) *nssIdentity {
+	return &nssIdentity{node: node}
+}
+
+func (nssStore) Import(data []byte, password string) error {
+	unicodePassword, err := bmpString(password)
+	if unicodePassword == nil {
+		return err
+	}
+	var pass = C.SECITEM_AllocItem(nil, nil, C.uint(len(unicodePassword)))
+	if pass == nil {
+		return errors.New("SECITEM_AllocItem failed")
+	}
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
+	var p12 = C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
+	var decoded = C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
+	if decoded != 0 {
+		return errors.New(fmt.Sprintf("Error %d, P12 decoding failed...\n", int(C.PR_GetError())))
+	}
+	var authenticated = C.SEC_PKCS12DecoderVerify(p12)
+	if authenticated != 0 {
+		return errors.New(fmt.Sprintf("Error %d, P12 authentication failed...\n", int(C.PR_GetError())))
+	}
+	var validated = C.SEC_PKCS12DecoderValidateBags(p12, (*[0]byte)(C.P12U_NicknameCollisionCallback))
+	if validated != 0 {
+		return errors.New(fmt.Sprintf("Error %d, P12 validation failed...\n", int(C.PR_GetError())))
+	}
+	var imported = C.SEC_PKCS12DecoderImportBags(p12)
+	if imported != 0 {
+		return errors.New(fmt.Sprintf("Error %d, P12 import failed...\n", int(C.PR_GetError())))
+	}
+	return nil
+}
+
+func (nssStore) Close() {
```

Comment:
We cannot, if one unit test will call close and it will call NSS_Shutdown then the app is done, calling NSS_InitReadWrite again will return error - thats by NSS design